crates/ruff_linter/src/rules/ruff/rules/class_dict_annotations.rs

```diff
@@ -0,0 +1,131 @@
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
+use ruff_macros::{ViolationMetadata, derive_message_formats};
+use ruff_python_ast::{Expr, ExprCall, PythonVersion};
+use ruff_text_size::Ranged;
+
+/// ## What it does
+/// Checks for uses of `<identifier>.__dict__.get("__annotations__" [, <default>])`
+/// on Python 3.10+ and Python < 3.10 with `typing_extensions` enabled.
+///
+/// ## Why is this bad?
+/// On Python 3.10 and newer, `<identifier>.__dict__.get("__annotations__" [, <default>])`
+/// can be unreliable, especially with the introduction of stringized annotations
+/// and features like `from __future__ import annotations`. For Python 3.14+, this
+/// approach is more strongly discouraged.
+///
+/// See [Python Annotations Best Practices](https://docs.python.org/3.14/howto/annotations.html)
+/// for alternatives.
+///
+/// TLDR:
+///
+/// If you use a class with a custom metaclass and access `__annotations__`
+/// on the class, you may observe unexpected behavior; see
+/// [PEP 749](https://peps.python.org/pep-0749/#pep749-metaclasses) for some
```

Comment:
This section is relevant to accessing `cls.__annotations__`, not `cls.__dict__.get("__annotations__")`. The former is unsafe in the presence of metaclasses (on 3.14+ only if `from __future__ import annotations` is used; on earlier versions always).

I think Ruff should have a separate rule against accessing `cls.__annotations__` on a class object, but that's not what this rule is about.

Reply:
I opened #18236 about this.

Reply:
This has been removed.